src/classlibnative/bcltype/arraynative.inl

```diff
@@ -0,0 +1,248 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+// File: ArrayNative.cpp
+//
+
+//
+// This file contains the native methods that support the Array class
+//
+
+#ifndef _ARRAYNATIVE_INL_
+#define _ARRAYNATIVE_INL_
+
+#include "gchelpers.inl"
+
+FORCEINLINE void InlinedForwardGCSafeCopyHelper(void *dest, const void *src, size_t len)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+        PRECONDITION(CheckPointer(dest));
+        PRECONDITION(CheckPointer(src));
+        SO_TOLERANT;
+    }
+    CONTRACTL_END;
+
+    _ASSERTE(dest != nullptr);
+    _ASSERTE(src != nullptr);
+    _ASSERTE(dest != src);
+    _ASSERTE(len != 0);
+
+    // To be able to copy forwards, the destination buffer cannot start inside the source buffer
+    _ASSERTE((SIZE_T)dest - (SIZE_T)src >= len);
+
+    // Make sure everything is pointer aligned
+    _ASSERTE(IS_ALIGNED(dest, sizeof(SIZE_T)));
+    _ASSERTE(IS_ALIGNED(src, sizeof(SIZE_T)));
+    _ASSERTE(IS_ALIGNED(len, sizeof(SIZE_T)));
+
+    _ASSERTE(CheckPointer(dest));
+    _ASSERTE(CheckPointer(src));
+
+    SIZE_T *dptr = (SIZE_T *)dest;
+    SIZE_T *sptr = (SIZE_T *)src;
+
+    if ((len & sizeof(SIZE_T)) != 0)
```

Comment:
And it is worth doing the alignment only on AMD64 (or other platforms where we will have custom implementations). It may look better to put the AMD64 implementation under one big ifdef.